src/cpu/aarch64/cpu_isa_traits.cpp

```diff
@@ -0,0 +1,196 @@
+// be asymptotically as expensive as a single lock-prefixed memory read. The
+// get() method also has a 'soft' mode when the setting is not locked for
+// re-setting. This is used for testing purposes.
+template <typename T>
+struct set_before_first_get_setting_t {
+private:
+    T value_;
+    bool initialized_;
+    std::atomic<unsigned> state_;
+    enum : unsigned { idle = 0, busy_setting = 1, locked_after_a_get = 2 };
+
+public:
+    set_before_first_get_setting_t(T init = T(0))
+        : value_ {init}, initialized_ {false}, state_ {0} {}
+
+    bool set(T new_value) {
+        if (state_.load() == locked_after_a_get) return false;
+
+        while (true) {
+            unsigned expected = idle;
+            if (state_.compare_exchange_weak(expected, busy_setting)) break;
+            if (expected == locked_after_a_get) return false;
+        }
+
+        value_ = new_value;
+        initialized_ = true;
+        state_.store(idle);
+        return true;
+    }
+
+    bool initialized() { return initialized_; }
+
+    T get(bool soft = false) {
+        if (!soft && state_.load() != locked_after_a_get) {
+            while (true) {
+                unsigned expected = idle;
+                if (state_.compare_exchange_weak(expected, locked_after_a_get))
+                    break;
+                if (expected == locked_after_a_get) break;
+            }
+        }
+        return value_;
+    }
+};
+
+set_before_first_get_setting_t<cpu_isa_t> &max_cpu_isa() {
+    static set_before_first_get_setting_t<cpu_isa_t> max_cpu_isa_setting;
+    return max_cpu_isa_setting;
+}
+
+bool init_max_cpu_isa() {
+    if (max_cpu_isa().initialized()) return false;
+
+    cpu_isa_t max_cpu_isa_val = isa_all;
+    char buf[64];
+    if (getenv("DNNL_MAX_CPU_ISA", buf, sizeof(buf)) > 0) {
+
+#define IF_HANDLE_CASE(cpu_isa) \
+    if (std::strcmp(buf, cpu_isa_traits<cpu_isa>::user_option_env) == 0) \
+    max_cpu_isa_val = cpu_isa
+#define ELSEIF_HANDLE_CASE(cpu_isa) else IF_HANDLE_CASE(cpu_isa)
+
+        IF_HANDLE_CASE(isa_all);
+        ELSEIF_HANDLE_CASE(simdfp);
+        ELSEIF_HANDLE_CASE(sve_512);
+
+#undef IF_HANDLE_CASE
+#undef ELSEIF_HANDLE_CASE
+    }
+
+    return max_cpu_isa().set(max_cpu_isa_val);
+}
+#endif
+} // namespace
+
+struct isa_info_t {
+    isa_info_t(cpu_isa_t aisa) : isa(aisa) {};
+
+    // this converter is needed as code base defines certain ISAs
+    // that the library does not expose (eg sve),
+    // so the internal and external enum types do not coincide.
+    dnnl_cpu_isa_t convert_to_public_enum(void) const {
+        switch (isa) {
+            case sve_512: return dnnl_cpu_isa_sve_512;
+            case simdfp: return dnnl_cpu_isa_simdfp;
+            default: return dnnl_cpu_isa_all;
+        }
+    }
+
+    const char *get_name() const {
+        switch (isa) {
+            case sve_512: return "AArch64 SVE (512 bits)";
+            case simdfp: return "AArch64 SIMD&FP";
+            default: return "AArch64";
```

Comment:
In fact, I didn't know that Advanced SIMD was mandatory for AArch64 :-), but I preserved `asimd`. https://github.com/oneapi-src/oneDNN/pull/850/commits/26a9f0a1b5d524f4abc1c5a748011bcf9d8f2038
If DNNL_MAX_CPU_ISA=ADVANCED_SIMD, only Advanced SIMD (no SVE) instructions are generated by JIT.
ADVANCED_SIMD (only Advanced SIMD instructions are used) < SVE_512 (both SVE_512 and Advanced SIMD instructions are used).
At the moment, oneDNN seems to no option to completely disable JIT. We should prepare this option?

Reply:
To me there are two things here: 1) disabling JIT and 2) limiting the maximum ISA the code can dispatch.

The question is: what are we trying to achieve?

For x64 the JIT is an important part of the library that essentially delivers the performance. There is no much sense in oneDNN on x64 w/o JIT. Hence, we don't have an option to opt-out from JIT.
However, we do allow to limit the maximum ISA that the library dispatches. There are two primary objectives here:
1. Allow users to debug the performance or correctness issues (strictly speaking for the latter part disabling jit could be useful).
2. We might disable dispatching some ISAs that are public, but for which there is no yet hardware that we validate oneDNN on. This is to prevent users that will use "old" (pre-silicon) oneDNN on future hardware from running into issues that we didn't see before running oneDNN or a real HW. However, today one may want to test how this future ISA would work or whatnot, so we give this `DNNL_MAX_CPU_ISA` option to allow opt-in for the new instructions. 
    - Example: see `AVX512_CORE_AMX` [here](https://github.com/oneapi-src/oneDNN/blob/master/doc/performance_considerations/dispatcher_control.md#run-time-controls)

It is worth noting that `DNNL_MAX_CPU_ISA` doesn't affect any libraries that oneDNN might rely on on x64. For instance, some time back oneDNN relied on Intel MKL, and using this option didn't affect the code dispatch on Intel MKL side, so users had to use `MKL_ENABLE_INSTRUCTION` together with `DNNL_MAX_CPU_ISA`. Luckily, currently oneDNN on x64 is completely self-contained, so the option truly controls oneDNN behavior.

So to me, since AArch64 implies ASIMD, there is no much sense in having it as a dedicated ISA. On the other hand, in imaginary world we might have 3 kinds of implementations: 1) Aarch64 scalar code (not using ASIMD), 2) Aarch64 code based on ASIMD, 3) Aarch64 code using SVE-xxx bit. And we might want to give a user an option to dispatch between those. In this case we might want to have a dedicated `Aarch64 + ASIMD` option.

So the long story short:
- Currently `DNNL_MAX_CPU_ISA` cannot be used to disable JIT. On x64 the "lowest" possible value is SSE4.1 which the minimum requirement for the library. So, no matter what, oneDNN will at least be able to use SSE4.1 JIT. `dnnl_cpu_isa_all` is used to remove all restrictions (see [here](https://github.com/oneapi-src/oneDNN/blob/65ec7af55fce9dae3176bda5bff690cddd8a33fb/include/dnnl_types.h#L2412-L2413))
    - We could add a special value for `DNNL_MAX_CPU_ISA` that would mean "please disable jit" (this value will have to be something like `0xffffffff`, since 0x0 is already taken by `any`). However it is unclear if this is really useful.
- If we don't think scalar code makes any sense, we probably should just say that Aarch64 (w/ ASIMD) is a bare minimum. The extensions come with Armv8.1-A, Armv8.2-A, SVE (various length), Armv-8.3-A, etc that user can optionally limit using `DNNL_MAX_CPU_ISA`.

Finally, if we don't have a clear picture on what we really want to do with ISA control on Aarch64 side, we could 1) call it experimental (so that we can adjust it in future once we have better understanding), or 2) not implementing the dispatcher limitation control at all (i.e. dispatching the maximum possible ISA regardless) and leave the implementation for later. The 2nd options is probably not the best one though :)

Thoughts?

Reply:
I think we have some complications on the AArch64 build that aren't an issue for x86.

What would we want from an 'AArch64 +ASIMD' option? Calling out 'AArch64 +ASIMD' as an ISA is, technically, meaningless, as discussed AdvancedSIMD is an inseparable part of AArch64 - so the compiler _will_ generate ASIMD instructions on v8 hardware. I guess this makes it the equivalent of `dnnl_cpu_isa_all` on Armv8-a.

I think what we're after is the ability to easily choose between `xbayak_aarch64`, `ACL`, and `C++` primitives? But the external Compute Library dependency (i.e. `-DDNNL_AARCH64_USE_ACL`) complicates things. ACL isn't bound by the ISA set in oneDNN and could be built to target `armv8-a+sve`.

Is the aim is to (eventually) be able to choose between the following options?
1. Xbayak_aarch64 SVE-<insert vlen here> (currently only SVE-512)
2. Xbayak_aarch64 ASIMD (currently no implementation, but there could be)
3. ACL SVE (currently not supported)
4. ACL ASIMD
5. None of the above, use ref. kernels.

As is stands, 3 & 4 are only possible when ACL is available as an external dependency and `DNNL_AARCH64_USE_ACL` is set, and the choice between 3 or 4 is down to how that external dependency was built. To pick each of those, I think you would need the `ACL` variants come first in the list `cpu_convolution_list` (for what is currently implemented), and then:

1. Build without ACL flags, for SVE-<insert vlen here> ISA.
2. Build without ACL flags, for 'all' ISA
3. Build with ACL flags, link to SVE build of libs
4. Build with ACL flags, link to non-SVE build of libs
If the `ACL` variants come first in the list in, for example, `cpu_convolution_list` then the choosing 3 or 4 is down to the the `DNNL_AARCH64_USE_ACL` option and how the ACL build `ACL_ROOT_DIR` points to was done.
5. ?

So there are two problems I see here:
1. Choice between 3 & 4 is out of oneDNN's control.
2. How to choose between 2 and 5 - this may be where we _do_ need to call out the ASIMD ISA? Even if that is slightly confusing, and maybe not technically correct! 

In short, I'm not convinced we _do_ have a clear picture of that the ISA control should be for AArch64, so I think we should consider whatever we do with this PR as experimental!

Reply:
> In short, I'm not entirely sure we _do_ have a clear picture of that the ISA control should be for AArch64, so I think we should consider whatever we do with this PR as experimental!

I second this judgment. Even on x64 we have roughly the same dilemma that you raised with the compiler (Intel's compiler can generate different versions of the function for a dedicated ISAs to have better performance at runtime, so this goes out of the control of the library), and with external libraries (even though on x64 this is less relevant, as typically no external libraries are used).

Ok, sounds we are aligned that we need to think through what we want and how we want to implement it, but that goes out of the scope of this PR. I am resolving this discussion then.

@kawakami-k, @nSircombe, thanks for sharing your thoughts!